src/lua/backtrace.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#pragma once
+
+#include "trivia/config.h"
+
+#ifdef ENABLE_BACKTRACE
+#include "core/backtrace.h"
+#include "core/say.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+enum {
+	/* +2 for `backtrace_collect` and `backtrace_collect_current_stack` */
+	/* Maximal length of a Lua function  name. */
+	BACKTRACE_LUA_PROC_NAME_LEN_MAX = 64,
+	/* Maximal length of a Lua source file name. */
+	BACKTRACE_LUA_SRC_NAME_LEN_MAX = 64,
+	/* Maximal number of Lua frames collected. */
+	BACKTRACE_LUA_FRAME_COUNT_MAX = 64,
+};
+
+/*
+ * Lua frame information, derived from backtrace frame.
+ */
+struct backtrace_lua_frame {
+	union {
+		/* Base frame type. */
+		struct backtrace_frame frame;
+		/* Derived frame type. */
+		struct {
+			/* Lua function name. */
+			char proc_name[BACKTRACE_LUA_PROC_NAME_LEN_MAX];
+			/* Lua source file name. */
+			char src_name[BACKTRACE_LUA_SRC_NAME_LEN_MAX];
+			/* Line number inside Lua function. */
+			int line_no;
+		};
+	};
+	enum backtrace_lua_frame_type {
+		BACKTRACE_FRAME,
+		BACKTRACE_LUA_FRAME,
```

Comment:
I suggest reworking this structure a bit to make the correspondence between types and data clearer:

```C
struct backtrace_lua_frame {
    union {
        struct backtrace_frame c;
        struct { ... } lua;
    };
    enum backtrace_lua_frame_type {
        BACKTRACE_LUA_FRAME_C,
        BACKTRACE_LUA_FRAME_LUA,
    };
};
```

Reply:
_Polymorphism_ looks quite ugly in C anyways...

Fixed: made the union members and enum elements conform to the corresponding language types.